Tasks/AzureCloudPowerShellDeployment/Publish-AzureCloudDeployment-Legacy.ps1

```diff
@@ -0,0 +1,296 @@
+
+                    Write-Host "New-AzureServiceDiagnosticsExtensionConfig -Role $role -StorageContext <context> -DiagnosticsConfigurationPath $fullExtPath"
+                    $wadconfig = New-AzureServiceDiagnosticsExtensionConfig -Role $role -StorageContext $storageContext -DiagnosticsConfigurationPath $fullExtPath 
+
+                    #Add each extension configuration to the array for use by caller
+                    $diagnosticsConfigurations += $wadconfig
+                }
+            }
+        }
+        else
+        {
+            Write-Warning (Get-LocalizedString -Key "Could not get the primary storage key for storage account '{0}'. Unable to apply any diagnostics extensions." -ArgumentList "$storageAccount")
+        }
+    }
+    
+    return $diagnosticsConfigurations
+}
+
+Write-Verbose "Entering script Publish-AzureCloudDeployment.ps1"
+
+import-module "Microsoft.TeamFoundation.DistributedTask.Task.Internal"
+import-module "Microsoft.TeamFoundation.DistributedTask.Task.Common"
+
+Write-Host "ConnectedServiceName= $ConnectedServiceName "
+Write-Host "ServiceName= $ServiceName"
+Write-Host "ServiceLocation= $ServiceLocation"
+Write-Host "AffinityGroup= $AffinityGroup"
+Write-Host "StorageAccount= $StorageAccount"
+Write-Host "CsPkg= $CsPkg"
+Write-Host "CsCfg= $CsCfg"
+Write-Host "Slot= $Slot"
+Write-Host "DeploymentLabel= $DeploymentLabel"
+Write-Host "AppendDateTimeToLabel= $AppendDateTimeToLabel"
+Write-Host "AllowUpgrade= $AllowUpgrade"
+Write-Host "NewServiceAdditionalArguments= $NewServiceAdditionalArguments"
+
+$allowUpgrade = Convert-String $AllowUpgrade Boolean
+
+Write-Host "Find-Files -SearchPattern $CsCfg"
+$serviceConfigFile = Find-Files -SearchPattern "$CsCfg"
+Write-Host "serviceConfigFile= $serviceConfigFile"
+$serviceConfigFile = Get-SingleFile $serviceConfigFile $CsCfg
+
+Write-Host "Find-Files -SearchPattern $CsPkg"
+$servicePackageFile = Find-Files -SearchPattern "$CsPkg"
+Write-Host "servicePackageFile= $servicePackageFile"
+$servicePackageFile = Get-SingleFile $servicePackageFile $CsPkg
+
+Write-Host "Get-AzureService -ServiceName $ServiceName -ErrorAction SilentlyContinue  -ErrorVariable azureServiceError"
+$azureService = Get-AzureService -ServiceName $ServiceName -ErrorAction SilentlyContinue  -ErrorVariable azureServiceError
+
+if($azureServiceError){
+   $azureServiceError | ForEach-Object { Write-Warning $_.Exception.ToString() }
+}   
+
+   
+if (!$azureService)
+{    
+    $azureService = "New-AzureService -ServiceName `"$ServiceName`""
+    if($NewServiceAffinityGroup) {
+        $azureService += " -AffinityGroup `"$NewServiceAffinityGroup`""
+    }
+    elseif($ServiceLocation) {
+         $azureService += " -Location `"$ServiceLocation`""
+    }
+    else {
+        throw "Either AffinityGroup or ServiceLocation must be specified"
+    }
+    $azureService += " $NewServiceAdditionalArguments"
+    Write-Host "$azureService"
+    $azureService = Invoke-Expression -Command $azureService
```

Comment:
You should build a hashtable and splat it instead of using Invoke-Expression which is subject to escaping fragilities.